[quant] Add quant APIs to save/load observer state_dict

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44846 [quant] Add quant APIs to save/load observer state_dict**

Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

supriyar committed:
[quant] Add quant APIs to save/load observer state_dict

Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
+    """
+    od = OrderedDict()
+    if isinstance(mod, torch.jit.RecursiveScriptModule):
+        for k, v in mod.state_dict().items():
```
would it be possible to return the state dict directly instead of copying it over?  it would reduce risk of bugs if there are any changes to how state dicts work and someone forgets to update this file

supriyar replied:
You mean update the original state_dict?
> it would reduce risk of bugs if there are any changes to how state dicts work and someone forgets to update this file

Wouldn't the unit tests catch this?

> would the current code copy the version over as well?
Not right now, but I can add that.

raghuramank100 replied:
Versioning is quite important. lets add support for that!

vkuzo replied:
currently we seem to be inspecting the state dict and copying things over one by one.  If instead we treat the state_dict for a module as a black box, then we remove the surface area of potential future bugs which could happen when state dict format changes (because we do not inspect it).

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
+    else:
+        return isinstance(module, torch.quantization.HistogramObserver)
+
+def _get_attr_names_for_obs(module, name):
```
would it work to load/save the state dict without inspecting it?  It would reduce the risk of bugs creeping up.

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
+
+    return od
+
+def load_observer_state_dict(mod, obs_dict):
```
instead of this, would it work to just load the state dict directly with strict==False, which will make it ignore missing keys for non-observer modules?  If that's possible, that would reduce the amount of custom code needed and eliminate another avenue for potential future bugs.

raghuramank100 replied:
We should also support a strict=False mode. Use case would be where we collect stats for all the modules, but quantize only a subset of them.

supriyar replied:
strict=False should already be supported. I can make sure the test covers that case too. For the case you mentioned it would still not print warning/error messages when the subset that you want to quantize is missing from the state_dict. 
As an alternate for this scenario can we load all stats, but modify convert to accept a qconfig that will utilize only a subset of them?

vkuzo replied:
I'm just concerned that we are inventing one more paradigm about how to save/load things, which can grow to be a lot of baggage over time.  Can we avoid this entirely by using current state dicts as is, and creating fb-only functions to handle things like missing observers, changed module names, etc?

supriyar replied:
I've updated the code to use a custom load_from_state_dict for PerChannelMinMaxObservers (since this observer requires resizing the min/max_vals to the right shape). For the rest it uses the default `load_from_state_dict`

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
+            isinstance(module, torch.quantization.FakeQuantize) or
+            _is_observer_script_module(module, 'torch.quantization.observer'))
+
+def _is_per_channel_obs_instance(module):
```
Can we infer this from the qscheme attribute instead of fixing the class name here?

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
+    in the model. The name should correspond to the attr name in the model
+    state_dict.
+    """
+    qual_names = {}
```
As a better engineering task, should we always a have a generic "get_stats" method for observerbase that returns the stats? This way, if we create a new observer, we wouldnt have to change this code.

vkuzo replied:
or, we can reuse the state_dict which already does this for us :)

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -429,53 +411,32 @@ def forward(self, x):
 
         # save state_dict of model
-        import io
+        obs_dict = torch.quantization.get_observer_state_dict(model)
```
Currently we seem to be doing this only for activation observers. As we move to QAT, we will need similar functionality, but we will need to consider weight fake-quants too. Is there a plan to add weight observers to the state dict?

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -649,10 +651,26 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                # call the `_load_from_state_dict` function defined module.py
+                if torch.jit.is_scripting():
+                    if name == 'min_vals':
+                        self.min_vals = val
```
can this be `self.min_vals.copy_(val)`, to preserve the dtype and device of the source?

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -649,10 +651,26 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                    if name == 'min_vals':
+                        self.min_vals = val
+                    else:
+                        self.max_vals = val
```
same

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -649,10 +651,26 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                                                                        missing_keys, unexpected_keys, error_msgs)
+
+    @torch.jit.export
+    def _load_from_state_dict_script(self, state_dict: Union[Dict[str, torch.Tensor], Dict[str, torch.Tensor]],
```
this seems to call `_load_from_state_dict`, just curious on what would break if this function wasn't there? I might be missing something

supriyar replied:
I noticed that without this, the call to `load_from_state_dict` was calling the one defined in module.py. Once we script the model we lose access to the overridden `load_from_state_dict` so we need to call a different fn here to get around it.

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -1092,6 +1110,68 @@ def forward(self, x):
+            _is_observer_script_module(module, "torch.quantization.observer.MovingAveragePerChannelMinMaxObserver")
+    return False
+
+def get_observer_state_dict(mod):
```
thoughts about making most of these FB-only, since the use case is pretty specific? Or are there OSS needs as well?

supriyar replied:
That may be true right now, but Raghu mentioned these will come in use when we develop the numeric suite and auto quantizer for OSS.

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quant APIs to save/load observer state_dict"


Summary:
The save function traverses the model state dict to pick out the observer stats
load function traverse the module hierarchy to load the state dict into module attributes depending on observer type

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_save_observer_state_dict

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23746821](https://our.internmc.facebook.com/intern/diff/D23746821)

[ghstack-poisoned]
